### PR TITLE
implement interface to manage custom hostnames

### PR DESCRIPTION
**Pull request by ElvinEfendi**
The PR introduces new functions to manage custom hostnames of a zone.

### CODE REVIEW

File: custom_hostname.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+// CustomHostnameSSL represents the SSL section in a given custom hostname.
+type CustomHostnameSSL struct {
+	Status      string `json:"status,omitempty"`
+	Method      string `json:"method,omitempty"`
+	Type        string `json:"type,omitempty"`
+	CnameTarget string `json:"cname_target,omitempty"`
+	CnameName   string `json:"cname_name,omitempty"`
+}
+
+type CustomMetadata map[string]interface{}
```
Comment (elithrar) [inline custom_hostname.go]: `CustomMetadata defines custom metadata for the hostname. This requires logic to be implemented by Cloudflare to act on the data provided.`

File: custom_hostname.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+// CustomHostname represents a custom hostname in a zone.
+type CustomHostname struct {
+	ID             string            `json:"id,omitempty"`
+	Hostname       string            `json:"hostname,omitempty"`
+	SSL            CustomHostnameSSL `json:"ssl,omitempty"`
+	CustomMetadata CustomMetadata    `json:"custom_metadata,omitempty"`
+}
+
+type CustomHostnameResponse struct {
```
Comment (elithrar) [inline custom_hostname.go]: `CustomHostNameResponse` represents a response from the Custom Hostnames endpoints.

File: custom_hostname.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	Result CustomHostname `json:"result"`
+	Response
+}
+
+type CustomHostnameListResponse struct {
```
Comment (elithrar) [inline custom_hostname.go]: as above

File: custom_hostname.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		return nil, errors.Wrap(err, errUnmarshalError)
+	}
+
+	return response, nil
+}
+
+// CustomHostnames lists custom hostnames for a given zone.
+//
+// API reference: https://api.cloudflare.com/#custom-hostname-for-a-zone-list-custom-hostnames
+func (api *API) CustomHostnames(zoneID string, page int, filter CustomHostname) ([]CustomHostname, ResultInfo, error) {
```
Comment (elithrar) [inline custom_hostname.go]: 👌🏻

File: custom_hostname.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	var response *CustomHostnameResponse
+	err = json.Unmarshal(res, &response)
+	if err != nil {
+		return nil, errors.Wrap(err, errUnmarshalError)
+	}
+
+	return response, nil
+}
+
+// CustomHostnames lists custom hostnames for a given zone.
```
Comment (elithrar) [inline custom_hostname.go]: the docs should describe what `page` and `filter` do, and what they default to (1 and no filter)

File: custom_hostname.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+	var response CustomHostnameResponse
+	err = json.Unmarshal(res, &response)
+	if err != nil {
+		return CustomHostname{}, errors.Wrap(err, errUnmarshalError)
+	}
+
+	return response.Result, nil
+}
+
+// CustomHostnameIDByName retrieves a custom hostname's ID from the hostname.
```
Comment (elithrar) [inline custom_hostname.go]: `CustomHostnameIDByName retrieves the ID for the given hostname`. (matches the name of the arg)

**Comment by ElvinEfendi** [comment]
Thanks for reviewing @elithrar. I addressed your comments, can you take another 👀 ?

**Comment by elithrar** [comment]
@ElvinEfendi Merged. Thanks again!

**Pull request merged**